File: src/2020/03.rb
map = File.open("../../data/2020/03.txt").read.strip.split

def count_trees map, dx, dy
  (0...map.size).step(dy).count do |y|
    x = ((y / dy) * dx) % map.first.size

    map[y][x] == "#"
  end
end

part_1 = count_trees(map, 3, 1)

puts part_1

part_2 = [[1, 1], [3, 1], [5, 1], [7, 1], [1, 2]].map do |dx, dy|
  count_trees map, dx, dy
end.inject(:*)

puts part_2

File: src/2018/03.py
import pathlib
from collections import defaultdict

from utils import *

data = pathlib.Path("../../data/2018/03.txt").read_text(encoding="utf-8")
lines = data.strip().split("\n")

lines = [ints(line) for line in lines]

fabric = defaultdict(int)
non_overlapped = set()
touched = defaultdict(set)

for claim_id, col, row, width, height in lines:
    non_overlapped.add(claim_id)

    overlapped = set()

    for w in range(width):
        for h in range(height):
            r, c = row + h, col + w

            fabric[r, c] += 1
            touched[r, c].add(claim_id)

            if len(touched[r, c]) > 1:
                overlapped.update(touched[r, c])

    if overlapped:
        for o in overlapped:
            non_overlapped.discard(o)


part_1 = len(list(filter(lambda c: c > 1, fabric.values())))

print(part_1)

part_2 = non_overlapped.pop()

print(part_2)
